File: week_2/day_08_stack/next_greater_follow_up_questions/nextSmallerElement.js
function nextSmaller(nums) {
    const stack = [];
    const res = new Array(nums.length).fill(-1);

    for (let i = 0; i < nums.length; i++) {
        while (stack.length && nums[i] < nums[stack[stack.length - 1]]) {
            const idx = stack.pop();
            res[idx] = nums[i];
        }
        stack.push(i);
    }
    return res;
}

// INPUT
console.log(nextSmaller([4, 5, 2, 10])); // [2, 2, -1, -1]

File: week_2/day_08_stack/next_greater_follow_up_questions/next_smaller_element.py
def nextSmaller(nums):
    stack = []
    res = [-1] * len(nums)

    for i in range(len(nums)):
        while stack and nums[i] < nums[stack[-1]]:
            idx = stack.pop()
            res[idx] = nums[i]
        stack.append(i)

    return res

# INPUT
nums = [4, 5, 2, 10]
print(nextSmaller(nums))  # Output: [2, 2, -1, -1]
